src/Analyser/NodeScopeResolver.php

```diff
@@ -1128,6 +1128,8 @@ private function processStmtNode(
 					$hasYield = $hasYield || $caseResult->hasYield();
 					$throwPoints = array_merge($throwPoints, $caseResult->getThrowPoints());
 					$branchScope = $scopeForBranches->filterByTruthyValue($condExpr);
+
+					$branchScope = $branchScope->assignExpression($stmt->cond, $branchScope->getType($caseNode->cond));
```

Comment:
I don't agree with this, `switch` uses `==` semantics. So if you have:

```php
switch ($foo) {
    case true: doFoo();
    break;
}
```

`$foo` doesn't have to be `true`, it can be any truthy-string for example...

The way to achieve this is to make PHPStan aware of these semantics in TypeSpecifier, to make it learn the whole table "Loose comparisons with ==" here https://www.php.net/manual/en/types.comparisons.php.

If you want better analysis now, switch to the `match` expression which uses `===` semantics and is a much better alternative.

Reply:
Thx for the hint, will look into it.

I also realized that same type switch already works like I expected